src/cmd/flux-mini.py

```diff
@@ -0,0 +1,335 @@
+        slot = create_slot("task", num_tasks, [core])
+        resource_section = slot
+
+    jobspec = {
+        "version": 1,
+        "resources": [resource_section],
+        "tasks": [
+            {
+                "command": command,
+                "slot": "task",
+                "count": task_count_dict,
+                "attributes": {},
+            }
+        ],
+        "attributes": {"system": {"cwd": os.getcwd(), "environment": environ}},
+    }
+    if walltime:
+        jobspec["attributes"]["system"]["duration"] = walltime
+
+    return jobspec
+
+
+def common_validate_args(args):
+    if args.nodes > args.ntasks:
+        raise ValueError("Number of nodes greater than the number of tasks")
+
+    if (
+        args.time
+        and re.match(r"^(\d+-)?\d+:\d+:\d+$", args.time) is None
+        and re.match(r"^\d+(:\d+)?$", args.time) is None
+    ):
+        raise ValueError(
+            "invalid time limit string format. "
+            "Acceptable formats include minutes[:seconds], [days-]hours:minutes:seconds"
+        )
+
+
+#  Convert walltime string to floating point seconds
+def walltime_to_duration(time):
+    if not time:
+        return None
+    p1 = re.compile(
+        r"^((?P<days>\d+)-)?"
+        r"(?P<hours>\d+):"
+        r"(?P<minutes>\d+):"
+        r"(?P<seconds>\d+)$"
+    )
+    p2 = re.compile(r"^(?P<minutes>\d+)(:(?P<seconds>\d+))?$")
+
+    t = 0.0
+    m = p2.search(time) or p1.search(time)
+    if m is None:
+        return None
+    vals = {k: float(v) for k, v in m.groupdict().items() if v is not None}
+    if "days" in vals:
+        t = t + vals["days"] * 60 * 60 * 24
+    if "hours" in vals:
+        t = t + vals["hours"] * 60 * 60
+    if "minutes" in vals:
+        t = t + vals["minutes"] * 60
+    if "seconds" in vals:
+        t = t + vals["seconds"]
+    if t == 0.0:
+        return None
+    return t
+
+
+def kv_list_split(s):
+    """
+    Split a key/value list with optional values, e.g. 'FOO,BAR=baz,...'
+    and return a dict.
+    """
+    return dict((a, b) for a, _, b in [x.partition("=") for x in s.split(",")])
+
+
+def common_get_parser():
+    """
+    Arg parser for subcommand - common options for run and submit
+    """
+    parser = argparse.ArgumentParser(add_help=False)
+    parser.add_argument(
+        "-N", "--nodes", type=int, default=0, help="Number of nodes to allocate"
+    )
+    parser.add_argument("-n", "--ntasks", type=int, help="Number of tasks to start")
+    parser.add_argument(
+        "-c",
+        "--cpus-per-task",
+        type=int,
+        default=1,
+        help="Number of cpus to allocate per task",
+    )
+    parser.add_argument(
+        "-t", "--time", help="Time limit: min[:sec] or [days-]hours:min:sec"
+    )
+    parser.add_argument(
+        "--export",
+        metavar="[ALL|NONE|VARS]",
+        action="append",
+        type=kv_list_split,
+        default=[{}],
+        help="List or specify environment variables to export",
+    )
```

Comment:
If you drop the `append` action and set the `default={}`, then you can remove the `exports = {...}` processing in `get_environment`.  The `args.export` dictionary will already be in the format you need.

All of that assumes `--export` will only be provided on the command line one time.  If that's a bad asummption, that ignore everything I said and what you have is :ok_hand: :+1:.

Reply:
I just copied what was in `flux jobspec srun` here.  It does seem like supporting multiple occurrences of `--setenv`  is a plus.  However, the option was a clone of srun's and we're not tied to that here, so maybe there is room for improvement?

Reply:
I would avoid adding options to `flux mini run` unless they are critical at this point. I'm not sure that `--setenv` is used anywhere at this time.